File: dataSync/InLong-versions/inlong-1.3.0-RC1/inlong-agent/agent-core/src/main/java/org/apache/inlong/agent/core/task/Task.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.inlong.agent.core.task;

import org.apache.inlong.agent.conf.JobProfile;
import org.apache.inlong.agent.plugin.Channel;
import org.apache.inlong.agent.plugin.Reader;
import org.apache.inlong.agent.plugin.Sink;

/**
 * task meta definition which contains reader -> channel -> sink and job config information
 */
public class Task {

    private final String taskId;
    private final Reader reader;
    private final Sink sink;
    private final Channel channel;
    private final JobProfile jobConf;

    public Task(String taskId, Reader reader, Sink sink, Channel channel,
        JobProfile jobConf) {
        this.reader = reader;
        this.sink = sink;
        this.taskId = taskId;
        this.channel = channel;
        this.jobConf = jobConf;
    }

    public boolean isReadFinished() {
        return reader.isFinished();
    }

    public String getTaskId() {
        return taskId;
    }

    public Reader getReader() {
        return reader;
    }

    public Sink getSink() {
        return sink;
    }

    public Channel getChannel() {
        return channel;
    }

    public void init() {
        this.channel.init(jobConf);
        this.sink.init(jobConf);
        this.reader.init(jobConf);
    }

    public void destroy() {
        this.reader.destroy();
        this.sink.destroy();
        this.channel.destroy();
    }
}

File: factory/datart/frontend/src/task.ts
/**
 * Datart
 *
 * Copyright 2021
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// organize-imports-ignore  polyfill/stable must in the first
import 'react-app-polyfill/stable';
import 'core-js/features/string/replace-all';
import { migrateWidgets } from 'app/migration/BoardConfig/migrateWidgets';
import { ChartDataRequestBuilder } from 'app/models/ChartDataRequestBuilder';
import {
  DataChart,
  ServerDashboard,
} from 'app/pages/DashBoardPage/pages/Board/slice/types';
import { getBoardChartRequests } from 'app/pages/DashBoardPage/utils';
import {
  getChartDataView,
  getDashBoardByResBoard,
  getDataChartsByServer,
} from 'app/pages/DashBoardPage/utils/board';
import { getWidgetMap } from 'app/pages/DashBoardPage/utils/widget';
import { ChartConfig } from 'app/types/ChartConfig';
import { ChartDetailConfigDTO } from 'app/types/ChartConfigDTO';
import { ChartDTO } from 'app/types/ChartDTO';
import { convertToChartDto } from 'app/utils/ChartDtoHelper';

// import 'core-js/stable/map';
// need polyfill [Object.values,Array.prototype.find,new Map]

/**
 * @param ''
 * @description 'server task 定时任务 调用'
 */
const getBoardQueryData = (dataStr: string) => {
  var data = JSON.parse(dataStr || '{}') as ServerDashboard;

  // const renderMode: VizRenderMode = 'schedule';
  const dashboard = getDashBoardByResBoard(data);
  const { datacharts, views: serverViews, widgets: serverWidgets } = data;

  const dataCharts: DataChart[] = getDataChartsByServer(
    datacharts,
    serverViews,
  );
  const migratedWidgets = migrateWidgets(serverWidgets, dashboard.config.type);
  const { widgetMap, wrappedDataCharts } = getWidgetMap(
    migratedWidgets, // TODO
    dataCharts,
    dashboard.config.type,
    serverViews,
  );

  const allDataCharts: DataChart[] = dataCharts.concat(wrappedDataCharts);
  const viewViews = getChartDataView(serverViews, allDataCharts);

  const viewMap = viewViews.reduce((obj, view) => {
    obj[view.id] = view;
    return obj;
  }, {});

  const dataChartMap = allDataCharts.reduce((obj, dataChart) => {
    obj[dataChart.id] = dataChart;
    return obj;
  }, {});
  let downloadParams = getBoardChartRequests({
    widgetMap,
    viewMap,
    dataChartMap,
  });
  let fileName = dashboard.name;
  return JSON.stringify({ downloadParams, fileName });
};

const getChartQueryData = (dataStr: string) => {
  // see  handleCreateDownloadDataTask
  const data: ChartDTO = JSON.parse(dataStr || '{}');

  const chartData = convertToChartDto(data);
  const dataConfig: ChartDetailConfigDTO = chartData.config;

  const chartConfig: ChartConfig = dataConfig.chartConfig as ChartConfig;
  const builder = new ChartDataRequestBuilder(
    {
      ...chartData.view,
      computedFields: chartData.config.computedFields,
    },
    chartConfig?.datas,
    chartConfig?.settings,
    {},
    false,
    dataConfig?.aggregation,
  );
  let downloadParams = [builder.build()];
  let fileName = data?.name || 'chart';
  return JSON.stringify({ downloadParams, fileName });
};
const getQueryData = (type: 'chart' | 'board', dataStr: string) => {
  if (type === 'board') {
    return getBoardQueryData(dataStr);
  } else {
    return getChartQueryData(dataStr);
  }
};
export default getQueryData;
